### PR TITLE
Added new Decoupled components and updated documentation

### PR DESCRIPTION
**Pull request by hutch31**
Added Crossbar, AsyncFifo, Serializer, and Credit modules

Added unit tests for new modules, and updated the README with component descriptions

### CODE REVIEW

File: src/main/scala/chisel/lib/dclib/package.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package chisel.lib
+
+import chisel3._
+
+package object dclib {
+  def bin2grey(x : UInt) : UInt = (x >> 1.U) ^ x
```
Comment (ekiwi) [inline src/main/scala/chisel/lib/dclib/package.scala]: We have gray code conversion in the Chisel3 utils library now: https://github.com/chipsalliance/chisel3/blob/master/src/main/scala/chisel3/util/GrayCode.scala

Comment (hutch31) [reply src/main/scala/chisel/lib/dclib/package.scala]: Ok I'll switch to the lib versions and test.  Also noticed that the checks were failing after submitting the request so will address those as well.

**Comment by hutch31** [comment]
Fixed code formatting and failing ECC tests

File: src/main/scala/chisel/lib/dclib/package.scala (hunk 1 above)
Comment (hutch31) [reply src/main/scala/chisel/lib/dclib/package.scala]: Changed to chisel utils version of GrayCode in 910f360 below

**Comment by schoeberl** [comment]
I would like to jump into the discussion of those components. I find this library great. As you move it from ip-contributions to the Chisel stdlib, there are name changes possible. In general, I think the term decoupled was not well chosen in Chisel. The term used in literature is ready/valid interface. Maybe you could rename your library to use that RV in the name.

Furthermore, Chisel inherits its naming style from Scala, which itself inherits from Java. Maybe you could change some of your snake_case names to CamelCase. This is highly recommended: https://www.chisel-lang.org/chisel3/docs/developers/style.html

**Comment by schoeberl** [comment]
Nice, that was fast :-)

What about s/DC/RV/?

**Comment by hutch31** [comment]
W.r.t. s/DC/RV, while Decoupled may not be the best name for the interface, it is consistently used within all the Chisel libs, and I think that being consistent with Chisel conventions is more important than matching the more commonly used external term but being inconsistent in the library.

**Comment by schoeberl** [comment]
@ekiwi OK to merge?

**Review by ekiwi — APPROVED** [review]
LGTM.

@hutch31: have you tested the AsyncFifo on a FPGA?

**Comment by hutch31** [comment]
@ekiwi The AsyncFifo has not gone through FPGA testing yet but should happen in the next few months

**Comment by schoeberl** [comment]
I've changed the code a bit to avoid deprecated warnings. Now we have a merge conflict. Could you please take a look?

**Comment by hutch31** [comment]
Will do.

**Pull request merged**